### PR TITLE
Experiment adding libhdf5

### PR DESCRIPTION
**Pull request by willGraham01**
Opening this PR so we can have an ongoing discussion and a record of it somewhere.

## NOT INTENDED FOR REVIEW OR MERGING ANYTIME SOON

Tagging #181

### CODE REVIEW

**Comment by willGraham01** [comment]
[Linux build fails](https://github.com/UCL/TDMS/actions/runs/3376839910/jobs/5605044417), and this is replicable on [my local machine](https://github.com/UCL/TDMS/files/9919584/hdf5_build_log_ubuntu.log).

Can't make much from the error message - looks like (for some reason) HDF5 is expecting a different version to what's installed? `H5check_version()` has [some documentation](https://docs.hdfgroup.org/hdf5/develop/group___h5.html#ga93525482e1168344f4c92470f99d88be) - it's meant to be a consistency check for things that are already compiled.

**Comment by samcunliffe** [comment]
> [Linux build fails](https://github.com/UCL/TDMS/actions/runs/3376839910/jobs/5605044417), and this is replicable on [my local machine](https://github.com/UCL/TDMS/files/9919584/hdf5_build_log_ubuntu.log).
> 
> Can't make much from the error message - looks like (for some reason) HDF5 is expecting a different version to what's installed? `H5check_version()` has [some documentation](https://docs.hdfgroup.org/hdf5/develop/group___h5.html#ga93525482e1168344f4c92470f99d88be) - it's meant to be a consistency check for things that are already compiled.

That's indeed strange. We had it working in a little sandbox toy experiment, right? On your machine too? So it seems our `CMakeLists.txt` or `FindHDF5.cmake` is to blame.

**Comment by giordano** [comment]
What happened is that CMake [correctly found system hdf5](https://github.com/UCL/TDMS/actions/runs/3376872335/jobs/5605117959#step:8:28) during the configuration stage:
```
-- Found HDF5: /usr/lib/x86_64-linux-gnu/hdf5/serial/libhdf5.so;/usr/lib/x86_64-linux-gnu/libpthread.so;/usr/lib/x86_64-linux-gnu/libsz.so;/usr/lib/x86_64-linux-gnu/libz.so;/usr/lib/x86_64-linux-gnu/libdl.so;/usr/lib/x86_64-linux-gnu/libm.so (found version "1.10.4") 
```
but then when compiling the program the linker [picks up matlab's own hdf5](https://github.com/UCL/TDMS/actions/runs/3376872335/jobs/5605117959#step:9:155):
```
/usr/bin/c++ -Werror -O3 -DNDEBUG CMakeFiles/tdms.dir/src/openandorder.cpp.o CMakeFiles/tdms.dir/src/fields/base.cpp.o CMakeFiles/tdms.dir/src/fields/electric.cpp.o CMakeFiles/tdms.dir/src/fields/magnetic.cpp.o CMakeFiles/tdms.dir/src/fields/split.cpp.o CMakeFiles/tdms.dir/src/fields/td_field_exporter_2d.cpp.o CMakeFiles/tdms.dir/src/argument_parser.cpp.o CMakeFiles/tdms.dir/src/array_init.cpp.o CMakeFiles/tdms.dir/src/arrays.cpp.o CMakeFiles/tdms.dir/src/dimensions.cpp.o CMakeFiles/tdms.dir/src/fdtd_grid_initialiser.cpp.o CMakeFiles/tdms.dir/src/grid_labels.cpp.o CMakeFiles/tdms.dir/src/interface.cpp.o CMakeFiles/tdms.dir/src/interpolate.cpp.o CMakeFiles/tdms.dir/src/iterator.cpp.o CMakeFiles/tdms.dir/src/interpolation_methods.cpp.o CMakeFiles/tdms.dir/src/matrix_collection.cpp.o CMakeFiles/tdms.dir/src/mesh_base.cpp.o CMakeFiles/tdms.dir/src/numerical_derivative.cpp.o CMakeFiles/tdms.dir/src/shapes.cpp.o CMakeFiles/tdms.dir/src/simulation_parameters.cpp.o CMakeFiles/tdms.dir/src/source.cpp.o CMakeFiles/tdms.dir/src/timer.cpp.o CMakeFiles/tdms.dir/src/utils.cpp.o CMakeFiles/tdms.dir/matlabio/matlabio.cpp.o -o tdms   -L/usr/local/MATLAB/R2022b/bin/glnxa64  -Wl,-rpath,/home/runner/work/TDMS/build:/usr/local/MATLAB/R2022b/bin/glnxa64: libtdms_lib.so /usr/lib/x86_64-linux-gnu/libfftw3.so -lmex -lmx -lmat /usr/lib/gcc/x86_64-linux-gnu/9/libgomp.so /usr/lib/x86_64-linux-gnu/libpthread.so _deps/spdlog-build/libspdlog.a -lpthread 
/usr/bin/ld: CMakeFiles/tdms.dir/src/openandorder.cpp.o: undefined reference to symbol 'H5check_version@@MWHDF51.8.12'
/usr/bin/ld: /usr/local/MATLAB/R2022b/bin/glnxa64/libhdf5-1.8.so.8: error adding symbols: DSO missing from command line
```

There are two problems:

* the target `tdms` isn't configured to explicitly link to the found HDF5 library, so the linker just picks up the first one it finds.  Adding
  ```cmake
  target_link_libraries(tdms ${HDF5_CXX_LIBRARIES})
  ```
  should do the trick
* since you're using the C++ API, you need to require the C++ component:
  ```cmake
  find_package(HDF5 REQUIRED COMPONENTS CXX)
  ```

With these two changes I'm able to build `tdms` correctly on this branch.

File: tdms/cmake/targets.cmake [hunk 1]
```diff
@@ -44,6 +45,7 @@ function(test_target)
             ${Matlab_MX_LIBRARY}
             ${Matlab_MAT_LIBRARY}
             ${LIBCXX_LIBRARY}
+            ${HDF5_CXX_LIBRARIES}
```
Comment (samcunliffe) [inline tdms/cmake/targets.cmake:48]: > * the target tdms isn't told to explicitly link to the HDF5 library found during configuration, so the linker just picks up the first one it finds. Adding
>
>         target_link_libraries(tdms ${HDF5_CXX_LIBRARIES})
>   should do the trick

Mosè's fix is also needed for the dynlib target (at least on my ARM64 Mac which tries to link the wrong HDF5 libraries).

**Comment by samcunliffe** [comment]
I am starting to _actually_ develop (for #181) over on [`181-hdf5-io`](https://github.com/UCL/TDMS/tree/181-hdf5-io).
And, in parallel, get HDF5 working in Windows on #188.

Let's keep this PR open for now.

**Comment by samcunliffe** [comment]
Also closing this (c.f. #193) since it can just as happily live in the list of closed issues.
I'll need to revisit #193 (or something similar) given #215.

**Pull request closed without merging**